akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSupervisionSpec.scala

```diff
@@ -0,0 +1,656 @@
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set(RequestOne))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(3)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+    }
+
+    "restart when Expand `extrapolate` throws" in new TestSetup(Seq(Expand(
+      (in: Int) ⇒ in,
+      (agg: Int) ⇒ if (agg == 2) throw TE else (agg, -math.abs(agg)),
+      restartingDecider))) {
+
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(1)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(3)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-3)))
+    }
+
+    "resume when Expand `extrapolate` throws" in new TestSetup(Seq(Expand(
+      (in: Int) ⇒ in,
+      (agg: Int) ⇒ if (agg == 2) throw TE else (agg, -math.abs(agg)),
+      resumingDecider))) {
+
+      pending // see blow FIXME
+
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(1)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne() // FIXME this throws IllegalStateException: Cannot pull while holding, only pushAndPull
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(3)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-3)))
+    }
+
+    "resume when onPull throws before pushing all buffered elements" in new TestSetup(Seq(
+      BufferingTestStage(resumingDecider))) {
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(1)
+      lastEvents() should be(Set(OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(2)))
+
+      downstream.requestOne()
+      // 3 => boom
+      // FIXME feels wrong? it should deliver OnNext(4)?
+      lastEvents() should be(Set(RequestOne))
```

Comment:
> Thanks, this is interesting. Do you agree that for a restart we could support it, because the state is supposed to be cleared, and a new pull from upstream makes sense?

Unfortunately, no, because a fresh Stage expects a pull as the first event (except detached stages).

> Just for my understanding. Let's say that this custom stage would emit an element downstream here as response to the exception instead of pulling. If a special Filter stage was installed afterwards to discard that kind of error elements everything would work? I'm somewhat confused of why we could not emulate something like that?

Probably it will not work, because the stage will fail the pull exactly the same way as before (all pulls look the same). But there is something we can try: if a pull fails, we can (as you suggested) try to pull the resumed stage again. That way there is no lost token, but the new pull may fail just as the previous. We have some options here:

 - hope that repeating the pull will solve the issue, so try it until we hit a retry limit, then fail
 - add a user some kind of hook were he can "heal" state. Or alternatively we turn the resume into a restart.